## NYPD Motor Vehicle Collision Data

### Overview

The Motor Vehicle Collisions Crash table contains details on the crash events. Each row represents a crash event. The Motor Vehicle Collisions data tables contain information from all police reported motor vehicle collisions in NYC. The dataset can be found by following this link: https://data.cityofnewyork.us/Public-Safety/NYPD-Motor-Vehicle-Collisions-Crashes/h9gi-nx95

The data dates from 2012 to the current day, with data being updated on a daily basis. At the time of this writing, there are 1.59 million rows, each row representing a crash event, and over 29 columns which represent date, time, borough, zip code, latitude, longitude, location, on and off street name, cross street name, number of persons injured, number of persons killed, number of pedestrians injured, number of pedestrians killed, number of cyclist injured, number of cyclist killed, number of motorist injured, number of motorist killed, contributing factors, vehicle type codes and collision ID.

Let's begin by bringing in a few libraries and then bring the first three million rows of NYPD Motor Vehicle Collision Data using pandas. Then we will take a look at the data itself, make a few diagnoses, and execute some data cleaning fixes to improve my dataset for analysis.

In [1]:
import numpy as np
import pandas as pd
import datetime as dt
import sys
from IPython.core.display import display, HTML

In [5]:
datanyc = pd.read_csv("https://data.cityofnewyork.us/resource/h9gi-nx95.csv?$limit=3000000")

And let's pull up the data dictionary supplied by the OpenData website for reference.

In [71]:
data_dict = pd.read_excel("https://data.cityofnewyork.us/api/views/h9gi-nx95/files/2e58023a-21a6-4c76-b9e8-0101bf7509ca?download=true&filename=MVCollisionsDataDictionary.xlsx",
                         sheet_name='Column Info')
data_dict

Data Dictionary - Column Information                     Unnamed: 1  \
0                            Table Name                    Column Name   
1                 MV-Collisions - Crash                      UNIQUE_ID   
2                 MV-Collisions - Crash                  ACCIDENT_DATE   
3                 MV-Collisions - Crash                  ACCIDENT_TIME   
4                 MV-Collisions - Crash                        BOROUGH   
5                 MV-Collisions - Crash                       ZIP CODE   
6                 MV-Collisions - Crash                       LATITUDE   
7                 MV-Collisions - Crash                      LONGITUDE   
8                 MV-Collisions - Crash                       LOCATION   
9                 MV-Collisions - Crash                 ON STREET NAME   
10                MV-Collisions - Crash              CROSS STREET NAME   
11                MV-Collisions - Crash                OFF STREET NAME   
12                MV-Collisions - Crash      NUMBER OF PERSONS INJURED   
13                MV-Collisions - Crash       NUMBER OF PERSONS KILLED   
14                MV-Collisions - Crash  NUMBER OF PEDESTRIANS INJURED   
15                MV-Collisions - Crash   NUMBER OF PEDESTRIANS KILLED   
16                MV-Collisions - Crash      NUMBER OF CYCLIST INJURED   
17                MV-Collisions - Crash       NUMBER OF CYCLIST KILLED   
18                MV-Collisions - Crash     NUMBER OF MOTORIST INJURED   
19                MV-Collisions - Crash      NUMBER OF MOTORIST KILLED   
20                MV-Collisions - Crash  CONTRIBUTING FACTOR VEHICLE 1   
21                MV-Collisions - Crash  CONTRIBUTING FACTOR VEHICLE 2   
22                MV-Collisions - Crash  CONTRIBUTING FACTOR VEHICLE 3   
23                MV-Collisions - Crash  CONTRIBUTING FACTOR VEHICLE 4   
24                MV-Collisions - Crash  CONTRIBUTING FACTOR VEHICLE 5   
25                MV-Collisions - Crash            VEHICLE TYPE CODE 1   
26                MV-Collisions - Crash            VEHICLE TYPE CODE 2   
27                MV-Collisions - Crash            VEHICLE TYPE CODE 3   
28                MV-Collisions - Crash            VEHICLE TYPE CODE 4   
29                MV-Collisions - Crash            VEHICLE TYPE CODE 5   
30              MV-Collisions - Vehicle                      UNIQUE_ID   
31              MV-Collisions - Vehicle                   COLLISION_ID   
32              MV-Collisions - Vehicle                  ACCIDENT_DATE   
33              MV-Collisions - Vehicle                  ACCIDENT_TIME   
34              MV-Collisions - Vehicle                 VEHICLE_NUMBER   
35              MV-Collisions - Vehicle             STATE_REGISTRATION   
36              MV-Collisions - Vehicle                   VEHICLE_TYPE   
37              MV-Collisions - Vehicle                   VEHICLE_MAKE   
38              MV-Collisions - Vehicle                  VEHICLE_MODEL   
39              MV-Collisions - Vehicle                   VEHICLE_YEAR   
40              MV-Collisions - Vehicle               TRAVEL_DIRECTION   
41              MV-Collisions - Vehicle                  VEH_OCCUPANTS   
42              MV-Collisions - Vehicle                     DRIVER_SEX   
43              MV-Collisions - Vehicle          DRIVER_LICENSE_STATUS   
44              MV-Collisions - Vehicle    DRIVER_LICENSE_JURISDICTION   
45              MV-Collisions - Vehicle               PRE_ACDNT_ACTION   
46              MV-Collisions - Vehicle                POINT_OF_IMPACT   
47              MV-Collisions - Vehicle                 VEHICLE_DAMAGE   
48              MV-Collisions - Vehicle               VEHICLE_DAMAGE_1   
49              MV-Collisions - Vehicle               VEHICLE_DAMAGE_2   
50              MV-Collisions - Vehicle               VEHICLE_DAMAGE_3   
51              MV-Collisions - Vehicle         PUBLIC_PROPERTY_DAMAGE   
52              MV-Collisions - Vehicle    PUBLIC_PROPERTY_DAMAGE_TYPE   
53     

Let's see the first few rows of the dataset. 

In [3]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 200)
datanyc.head()

date   time    borough zip_code   latitude  longitude  \
0  2019-08-08T00:00:00.000  20:27  MANHATTAN    10033  40.848118 -73.930890   
1  2019-08-30T00:00:00.000  19:00        NaN      NaN  40.743214 -73.951350   
2  2019-08-18T00:00:00.000  17:56        NaN      NaN  40.839725 -73.783700   
3  2019-09-02T00:00:00.000   1:00   BROOKLYN    11208  40.670254 -73.857280   
4  2019-08-08T00:00:00.000  14:30        NaN      NaN  40.728603 -74.005325   

                       location                    on_street_name  \
0   POINT (-73.93089 40.848118)  AMSTERDAM AVENUE                   
1   POINT (-73.95135 40.743214)  PULASKI BRIDGE                     
2    POINT (-73.7837 40.839725)  CITY ISLAND AVENUE                 
3   POINT (-73.85728 40.670254)  LINDEN BOULEVARD                   
4  POINT (-74.005325 40.728603)  WEST HOUSTON STREET                

   off_street_name cross_street_name  number_of_persons_injured  \
0  WEST 181 STREET               NaN                        0.0   
1   JACKSON AVENUE               NaN                        0.0   
2    HORTON STREET               NaN                        1.0   
3  SAPPHIRE STREET               NaN                        0.0   
4              NaN               NaN                        0.0   

   number_of_persons_killed  number_of_pedestrians_injured  \
0                       0.0                              0   
1                       0.0                              0   
2                       0.0                              0   
3                       0.0                              0   
4                       0.0                              0   

   number_of_pedestrians_killed  number_of_cyclist_injured  \
0                             0                          0   
1                             0                          0   
2                             0                          0   
3                             0                          0   
4                             0                          0   

   number_of_cyclist_killed  number_of_motorist_injured  \
0                         0                           0   
1                         0                           0   
2                         0                           1   
3                         0                           0   
4                         0                           0   

   number_of_motorist_killed   contributing_factor_vehicle_1  \
0                          0  Driver Inattention/Distraction   
1                          0   Failure to Yield Right-of-Way   
2                          0                     Unspecified   
3                          0   Failure to Yield Right-of-Way   
4                          0  Driver Inattention/Distraction   

    contributing_factor_vehicle_2 contributing_factor_vehicle_3  \
0  Driver Inattention/Distraction                           NaN   
1                     Unspecified                           NaN   
2                     Unspecified                           NaN   
3                     Unspecified                           NaN   
4                     Unspecified                           NaN   

  contributing_factor_vehicle_4 contributing_factor_vehicle_5  unique_key  \
0                           NaN                           NaN     4185696   
1                           NaN                           NaN     4197471   
2                           NaN                           NaN     4190590   
3                           NaN                           NaN     4223885   
4                           NaN                           NaN     4185430   

  vehicle_type_code1                   vehicle_type_code2 vehicle_type_code_3  \
0              Sedan  Station Wagon/Sport Utility Vehicle                 NaN   
1              Sedan                                Sedan                 NaN   
2              Sedan  Station Wagon/Sport Utility Vehicle                 NaN   
3              Sedan                                 

... And get some overall information about the contents of the data.

In [4]:
pd.options.display.max_info_rows = 5000000
datanyc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1595863 entries, 0 to 1595862
Data columns (total 29 columns):
date                             1595863 non-null object
time                             1595863 non-null object
borough                          1112240 non-null object
zip_code                         1112049 non-null object
latitude                         1400333 non-null float64
longitude                        1400333 non-null float64
location                         1400333 non-null object
on_street_name                   1283621 non-null object
off_street_name                  1063439 non-null object
cross_street_name                220077 non-null object
number_of_persons_injured        1595846 non-null float64
number_of_persons_killed         1595832 non-null float64
number_of_pedestrians_injured    1595863 non-null int64
number_of_pedestrians_killed     1595863 non-null int64
number_of_cyclist_injured        1595863 non-null int64
number_of_cyclist_killed        

There are some missing values. Let's find the percentage of the missing values and see which columns have the most amount of missing values. We will get a mean of missing values and then round it to 4 decimal place.

In [6]:
pd.set_option('display.max_columns', 29)
datanyc.isnull().mean().round(4) * 100

date                              0.00
time                              0.00
borough                          30.30
zip_code                         30.32
latitude                         12.25
longitude                        12.25
location                         12.25
on_street_name                   19.57
off_street_name                  33.36
cross_street_name                86.21
number_of_persons_injured         0.00
number_of_persons_killed          0.00
number_of_pedestrians_injured     0.00
number_of_pedestrians_killed      0.00
number_of_cyclist_injured         0.00
number_of_cyclist_killed          0.00
number_of_motorist_injured        0.00
number_of_motorist_killed         0.00
contributing_factor_vehicle_1     0.27
contributing_factor_vehicle_2    13.43
contributing_factor_vehicle_3    93.54
contributing_factor_vehicle_4    98.66
contributing_factor_vehicle_5    99.66
unique_key                        0.00
vehicle_type_code1                0.34
vehicle_type_code2       

Wow some columns have a lot of missing values. For some it makes sense. For example contributing_factor_vehicle_2 or contributing_factor_vehicle_3 may be missing simply because there were no second or third vehicle contributing to the collision. Other than those, for example 'borough' column has 30% of its values missing. 

It looks like contributing_factor_vehicle_3, contributing_factor_vehicle_5, vehicle_type_code_3 and vehicle_type_code_5 have very few values, compared to the others. Lets take a aquick peek.

In [16]:
datanyc['contributing_factor_vehicle_1'].value_counts()

Unspecified                                              587958
Driver Inattention/Distraction                           296063
Failure to Yield Right-of-Way                             90524
Following Too Closely                                     78979
Backing Unsafely                                          61010
Other Vehicular                                           50754
Fatigued/Drowsy                                           46847
Turning Improperly                                        40835
Passing or Lane Usage Improper                            37277
Passing Too Closely                                       34169
Unsafe Lane Changing                                      30302
Traffic Control Disregarded                               24341
Driver Inexperience                                       23430
Lost Consciousness                                        19618
Pavement Slippery                                         15788
Prescription Medication                 

In [17]:
datanyc['contributing_factor_vehicle_3'].value_counts()

Unspecified                                              95922
Other Vehicular                                           1901
Driver Inattention/Distraction                            1376
Following Too Closely                                     1250
Fatigued/Drowsy                                            853
Pavement Slippery                                          290
Outside Car Distraction                                    147
Driver Inexperience                                        143
Reaction to Uninvolved Vehicle                             143
Traffic Control Disregarded                                137
Failure to Yield Right-of-Way                              120
Backing Unsafely                                            86
Unsafe Speed                                                77
Alcohol Involvement                                         76
Passing or Lane Usage Improper                              75
Turning Improperly                                     

We'll use the describe function which will "generate descriptive statistics that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding NaN values." This will work on numeric and object series, and may point out any glaring wholes in the data.

In [14]:
datanyc.describe()

latitude     longitude  number_of_persons_injured  \
count  1.400333e+06  1.400333e+06               1.595846e+06   
mean   4.069437e+01 -7.387704e+01               2.624765e-01   
std    1.099378e+00  2.284379e+00               6.599610e-01   
min    0.000000e+00 -2.013600e+02               0.000000e+00   
25%    4.066880e+01 -7.397725e+01               0.000000e+00   
50%    4.072258e+01 -7.392987e+01               0.000000e+00   
75%    4.076796e+01 -7.386705e+01               0.000000e+00   
max    4.334444e+01  0.000000e+00               4.300000e+01   

       number_of_persons_killed  number_of_pedestrians_injured  \
count              1.595832e+06                   1.595863e+06   
mean               1.167416e-03                   5.046611e-02   
std                3.612709e-02                   2.314794e-01   
min                0.000000e+00                   0.000000e+00   
25%                0.000000e+00                   0.000000e+00   
50%                0.000000e+00                   0.000000e+00   
75%                0.000000e+00                   0.000000e+00   
max                8.000000e+00                   2.700000e+01   

       number_of_pedestrians_killed  number_of_cyclist_injured  \
count                  1.595863e+06               1.595863e+06   
mean                   6.297533e-04               2.082510e-02   
std                    2.567946e-02               1.439698e-01   
min                    0.000000e+00               0.000000e+00   
25%                    0.000000e+00               0.000000e+00   
50%                    0.000000e+00               0.000000e+00   
75%                    0.000000e+00               0.000000e+00   
max                    6.000000e+00               4.000000e+00   

       number_of_cyclist_killed  number_of_motorist_injured  \
count              1.595863e+06                1.595863e+06   
mean               8.396711e-05                1.912852e-01   
std                9.231108e-03                6.222734e-01   
min                0.000000e+00                0.000000e+00   
25%                0.000000e+00                0.000000e+00   
50%                0.000000e+00                0.000000e+00   
75%                0.000000e+00                0.000000e+00   
max                2.000000e+00                4.300000e+01   

       number_of_motorist_killed    unique_key  
count               1.595863e+06  1.595863e+06  
mean                4.549263e-04  2.765198e+06  
std                 2.320963e-02  1.505727e+06  
min                 0.000000e+00  2.200000e+01  
25%                 0.000000e+00  1.026416e+06  
50%                 0.000000e+00  3.434054e+06  
75%                 0.000000e+00  3.833252e+06  
max                 5.000000e+00  4.233029e+06

Obviously averages and standard deviations don't tell a lot about latitude and longitude, but why we don't have full data counts in those columns. The rest of analysis doesn't appear to have obvious problems.

Let's take a look at some of those rows.

In [15]:
datanyc[datanyc['latitude'].isnull()]

date   time        borough zip_code  latitude  \
25       2019-08-04T00:00:00.000   4:01            NaN      NaN       NaN   
36       2019-08-29T00:00:00.000  15:34            NaN      NaN       NaN   
57       2019-08-10T00:00:00.000   3:45            NaN      NaN       NaN   
62       2019-08-13T00:00:00.000   9:30            NaN      NaN       NaN   
71       2019-08-24T00:00:00.000  23:50            NaN      NaN       NaN   
88       2019-08-22T00:00:00.000  16:36            NaN      NaN       NaN   
90       2019-08-16T00:00:00.000  11:20            NaN      NaN       NaN   
96       2019-09-01T00:00:00.000  17:03            NaN      NaN       NaN   
97       2019-08-23T00:00:00.000  18:47            NaN      NaN       NaN   
98       2019-08-09T00:00:00.000  10:00            NaN      NaN       NaN   
103      2019-08-02T00:00:00.000  15:00            NaN      NaN       NaN   
116      2019-08-10T00:00:00.000  11:00            NaN      NaN       NaN   
121      2019-08-08T00:00:00.000  12:28            NaN      NaN       NaN   
129      2019-10-29T00:00:00.000  13:55            NaN      NaN       NaN   
154      2019-08-28T00:00:00.000   0:13            NaN      NaN       NaN   
155      2019-10-29T00:00:00.000  15:59            NaN      NaN       NaN   
156      2019-08-30T00:00:00.000  17:30            NaN      NaN       NaN   
159      2019-08-23T00:00:00.000   0:45            NaN      NaN       NaN   
172      2019-08-22T00:00:00.000  20:00            NaN      NaN       NaN   
177      2019-08-29T00:00:00.000  19:00            NaN      NaN       NaN   
182      2019-10-29T00:00:00.000  16:40            NaN      NaN       NaN   
192      2019-08-24T00:00:00.000  17:50         QUEENS    11697       NaN   
197      2019-08-27T00:00:00.000  13:15            NaN      NaN       NaN   
199      2019-08-20T00:00:00.000  15:00            NaN      NaN       NaN   
204      2019-08-30T00:00:00.000  14:50         QUEENS    11355       NaN   
210      2019-10-29T00:00:00.000  17:32            NaN      NaN       NaN   
229      2019-09-03T00:00:00.000   9:10            NaN      NaN       NaN   
230      2019-08-27T00:00:00.000  13:30            NaN      NaN       NaN   
281      2019-08-06T00:00:00.000   9:45            NaN      NaN       NaN   
287      2019-10-29T00:00:00.000  19:41            NaN      NaN       NaN   
308      2019-08-06T00:00:00.000   9:08      MANHATTAN    10019       NaN   
309      2019-08-30T00:00:00.000  11:15            NaN      NaN       NaN   
310      2019-08-12T00:00:00.000  12:45            NaN      NaN       NaN   
314      2016-03-26T00:00:00.000  15:19      MANHATTAN    10033       NaN   
315      2019-10-29T00:00:00.000  20:06            NaN      NaN       NaN   
347      2019-08-27T00:00:00.000  23:45            NaN      NaN       NaN   
350      2019-08-08T00:00:00.000  18:00            NaN      NaN       NaN   
362      2019-08-24T00:00:00.000  14:40            NaN      NaN       NaN   
368      2019-10-29T00:00:00.000   2:30            NaN      NaN       NaN   
379      2019-08-23T00:00:00.000  22:52            NaN      NaN       NaN   
384      2019-09-01T00:00:00.000  14:30            NaN      NaN       NaN   
397      2019-08-15T00:00:00.000  23:50            NaN      NaN       NaN   
400      2019-08-16T00:00:00.000   0:48            NaN      NaN       NaN   
405      2019-08-12T00:00:00.000  17:03       BROOKLYN    11239       NaN   
421      2019-10-29T00:00:00.000   7:05            NaN      NaN       NaN   
445      2019-08-10T00:00:00.000  14:00          BRONX    10475       NaN   
447      2019-08-17T00:00:00.000  21:30            NaN      NaN       NaN   
454      2019-08-25T00:00:00.000   4:28            NaN      NaN       NaN   
495      2019-08-06T00:00:00.000   9:00            NaN      NaN       NaN   
504      2019-08-14T00:00:00.000  15:10            NaN      NaN       NaN   
513      2019-08-06T00:00:00.000  19:05            NaN      NaN       NaN   
525      2019-09-02T00:0

Seems like those rows have valid data. We wouldn't delete these rows, as the injury/fatality data may be useful, but we would drop them for location purposes.

If we really had some time, we would write/find a program to reverse map the on_street_name to the other fields.

Let's take a closer look at vehicle types.

In [18]:
datanyc['vehicle_type_code1'].value_counts()

PASSENGER VEHICLE                      715236
SPORT UTILITY / STATION WAGON          313500
Sedan                                  151360
Station Wagon/Sport Utility Vehicle    122291
TAXI                                    50670
VAN                                     26540
OTHER                                   23982
PICK-UP TRUCK                           23069
UNKNOWN                                 19929
Taxi                                    15635
SMALL COM VEH(4 TIRES)                  14559
LARGE COM VEH(6 OR MORE TIRES)          14527
BUS                                     14057
LIVERY VEHICLE                          10481
Pick-up Truck                           10168
Box Truck                                7965
Bus                                      6555
MOTORCYCLE                               6536
BICYCLE                                  5568
Bike                                     3916
Tractor Truck Diesel                     3496
Van                               

In [19]:
datanyc['vehicle_type_code_3'].value_counts()

PASSENGER VEHICLE                      63655
SPORT UTILITY / STATION WAGON          33161
Sedan                                  10248
Station Wagon/Sport Utility Vehicle     8477
UNKNOWN                                 3285
TAXI                                    3218
PICK-UP TRUCK                           2292
VAN                                     1489
OTHER                                   1108
Taxi                                     649
Pick-up Truck                            575
BICYCLE                                  533
SMALL COM VEH(4 TIRES)                   479
MOTORCYCLE                               464
LARGE COM VEH(6 OR MORE TIRES)           448
LIVERY VEHICLE                           424
BUS                                      403
Box Truck                                192
Bus                                      118
Motorcycle                                98
Bike                                      90
Tractor Truck Diesel                      87
Van       

We will leave this data for now, but modify the column names to standardize the dataset using the rename function.

## What Needs Attention?

#### STEP 1
* Some columns (such as vehicle_type_code_4, contributing_factor_vehicle_5) are nearly entirely empty. We'll remove those. 
* We will not be using some columns (e.g. unique_key, on_street_name, off_street_name, cross_street_name) so we can drop them completely. 
* We will change the data type of zip code to string.

#### STEP 2
* Cleaning or combining duplicated rows
* Renaming some columns
* Correcting misspellings
* Dealing with missing values
* Some dtype changes

#### STEP 3
* 'Latitude' and 'longitude' columns seem to be contained in the 'location' column. We like keeping the two values separate for now, so we can probably remove 'location' later.

#### STEP 4
* The values that we expect to be 'datetime' type are object ('date' and 'time' columns). We'll fix those.

#### STEP 5
* We are curious to see if collisions go up seasonally, so we'll make a new variable that bins the collisions by Spring (March, April, May), Summer (June, July, August), Fall (September, October, November), and Winter (December, January, February).

#### STEP 6
* We will do some general analyses to understand more about the data set.

Let's start! Just in case we mess anything up, we'll do all our cleaning in a new DataFrame called clean_nyc.

## Step 1: Dropping some columns and changing dtype
We'll begin by removing some columns, keeping only those missing fewer than 30% of their values. We'll also change the zip_code to a string. We can also drop some columns we know we're not going to use. Those operations are simple enough that we'll do them all before checking in again on the DataFrame.

In [7]:
clean_nyc = datanyc.dropna(thresh=(0.30 * datanyc.shape[0]), axis=1).copy()

In [8]:
clean_nyc.loc[:,'zip_code'] = clean_nyc['zip_code'].astype(str)

In [9]:
clean_nyc.drop(columns=["unique_key", "on_street_name", "off_street_name"], inplace=True)

Let's take a peek at what clean_nyc looks like now, as far as data types and number of columns (and values in those columns):

In [10]:
clean_nyc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1595863 entries, 0 to 1595862
Data columns (total 19 columns):
date                             1595863 non-null object
time                             1595863 non-null object
borough                          1112240 non-null object
zip_code                         1595863 non-null object
latitude                         1400333 non-null float64
longitude                        1400333 non-null float64
location                         1400333 non-null object
number_of_persons_injured        1595846 non-null float64
number_of_persons_killed         1595832 non-null float64
number_of_pedestrians_injured    1595863 non-null int64
number_of_pedestrians_killed     1595863 non-null int64
number_of_cyclist_injured        1595863 non-null int64
number_of_cyclist_killed         1595863 non-null int64
number_of_motorist_injured       1595863 non-null int64
number_of_motorist_killed        1595863 non-null int64
contributing_factor_vehicle_1    1

And what about the percentage of the missing values now?

In [11]:
pd.set_option('display.max_columns', 29)
clean_nyc.isnull().mean().round(4) * 100

date                              0.00
time                              0.00
borough                          30.30
zip_code                          0.00
latitude                         12.25
longitude                        12.25
location                         12.25
number_of_persons_injured         0.00
number_of_persons_killed          0.00
number_of_pedestrians_injured     0.00
number_of_pedestrians_killed      0.00
number_of_cyclist_injured         0.00
number_of_cyclist_killed          0.00
number_of_motorist_injured        0.00
number_of_motorist_killed         0.00
contributing_factor_vehicle_1     0.27
contributing_factor_vehicle_2    13.43
vehicle_type_code1                0.34
vehicle_type_code2               16.46
dtype: float64

So far, so good.

## Step 2: Getting Rid of Duplicated Rows, Correcting Misspellings and Renaming

We will modify the column names to standardize the dataset using the rename function.

In [26]:
datanyc.rename(columns={'vehicle_type_code1':'vehicle_type_code_1',
                        'vehicle_type_code2':'vehicle_type_code_2',
                       }, 
               inplace=True)

Let's take a closer look at vehicle_type_code_1.

In [59]:
datanyc['vehicle_type_code_1'].value_counts()

PASSENGER VEHICLE                      715236
Station Wagon/Sport Utility Vehicle    435791
Sedan                                  151360
Taxi                                    66305
Pick-up Truck                           33237
Van                                     29425
OTHER                                   23982
Bus                                     20612
UNKNOWN                                 19929
SMALL COM VEH(4 TIRES)                  14559
LARGE COM VEH(6 OR MORE TIRES)          14527
LIVERY VEHICLE                          10481
Box Truck                                7967
MOTORCYCLE                               6536
BICYCLE                                  5568
Bike                                     3916
Ambulance                                3699
Tractor Truck Diesel                     3496
TK                                       2485
BU                                       2229
Motorcycle                               1958
Convertible                       

'Sports utility/station wagon', 'sedan', and then 'station wagon/sport utility' were the largest number of vehicles involved in the crashes. We could combine 'sports utility/station wagon' and 'station wagon/sport utility' groups so that 'sport utility / station wagon' and 'station wagon/sport utility' vehicles can be under 'SUV'.

In [61]:
datanyc['vehicle_type_code_1'] = datanyc['vehicle_type_code_1'].str.replace('SPORT UTILITY / STATION WAGON', 'SUV')
datanyc['vehicle_type_code_1'] = datanyc['vehicle_type_code_1'].str.replace('Station Wagon/Sport Utility Vehicle', 'SUV')
datanyc['vehicle_type_code_1'].value_counts()

PASSENGER VEHICLE                 715236
SUV                               435797
Sedan                             151360
Taxi                               66305
Pick-up Truck                      33237
Van                                29425
OTHER                              23982
Bus                                20612
UNKNOWN                            19929
SMALL COM VEH(4 TIRES)             14559
LARGE COM VEH(6 OR MORE TIRES)     14527
LIVERY VEHICLE                     10481
Box Truck                           7967
MOTORCYCLE                          6536
BICYCLE                             5568
Bike                                3916
Ambulance                           3699
Tractor Truck Diesel                3496
TK                                  2485
BU                                  2229
Motorcycle                          1958
Convertible                         1333
Dump                                1190
FIRE TRUCK                          1065
DS              

When looking at value_counts() of contributing factors, we noted some duplicates. Let's clean those up now, first by looking at the unique values.

In [21]:
datanyc['contributing_factor_vehicle_1'].unique()

array(['Driver Inattention/Distraction', 'Failure to Yield Right-of-Way',
       'Unspecified', 'Turning Improperly', 'Traffic Control Disregarded',
       'Following Too Closely', 'Unsafe Speed',
       'Passing or Lane Usage Improper', 'Backing Unsafely',
       'Obstruction/Debris', 'Driver Inexperience', 'Alcohol Involvement',
       'Other Vehicular', 'Passing Too Closely', 'Fatigued/Drowsy',
       'View Obstructed/Limited', 'Aggressive Driving/Road Rage',
       'Reaction to Uninvolved Vehicle',
       'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion',
       'Unsafe Lane Changing', 'Passenger Distraction',
       'Oversized Vehicle', 'Glare', nan, 'Brakes Defective',
       'Pavement Slippery', 'Pavement Defective', 'Failure to Keep Right',
       'Steering Failure', 'Outside Car Distraction',
       'Lost Consciousness', 'Fell Asleep', 'Accelerator Defective',
       'Tire Failure/Inadequate', 'Drugs (illegal)', 'Animals Action',
       'Traffic Control Device Improper/N

There are some misspellings (e.g. "Illnes" missspelled in addition to 'Illness', in 'Drugs (illegal)' and 'Drugs (Illegal)'). Let's fix those.

In [22]:
datanyc['contributing_factor_vehicle_1'].replace('Illnes', 'Illness', inplace=True)
datanyc['contributing_factor_vehicle_1'].replace('Drugs (illegal)', 'Drugs (Illegal)', inplace=True)
datanyc['contributing_factor_vehicle_1'].unique()

array(['Driver Inattention/Distraction', 'Failure to Yield Right-of-Way',
       'Unspecified', 'Turning Improperly', 'Traffic Control Disregarded',
       'Following Too Closely', 'Unsafe Speed',
       'Passing or Lane Usage Improper', 'Backing Unsafely',
       'Obstruction/Debris', 'Driver Inexperience', 'Alcohol Involvement',
       'Other Vehicular', 'Passing Too Closely', 'Fatigued/Drowsy',
       'View Obstructed/Limited', 'Aggressive Driving/Road Rage',
       'Reaction to Uninvolved Vehicle',
       'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion',
       'Unsafe Lane Changing', 'Passenger Distraction',
       'Oversized Vehicle', 'Glare', nan, 'Brakes Defective',
       'Pavement Slippery', 'Pavement Defective', 'Failure to Keep Right',
       'Steering Failure', 'Outside Car Distraction',
       'Lost Consciousness', 'Fell Asleep', 'Accelerator Defective',
       'Tire Failure/Inadequate', 'Drugs (Illegal)', 'Animals Action',
       'Traffic Control Device Improper/N

In [23]:
datanyc['contributing_factor_vehicle_2'].unique()

array(['Driver Inattention/Distraction', 'Unspecified', nan,
       'Following Too Closely', 'Passing or Lane Usage Improper',
       'Passing Too Closely', 'Failure to Yield Right-of-Way',
       'Aggressive Driving/Road Rage',
       'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion', 'Glare',
       'Unsafe Speed', 'Backing Unsafely', 'Failure to Keep Right',
       'View Obstructed/Limited', 'Reaction to Uninvolved Vehicle',
       'Other Vehicular', 'Driver Inexperience', 'Turning Improperly',
       'Traffic Control Disregarded', 'Unsafe Lane Changing',
       'Obstruction/Debris', 'Pavement Slippery', 'Oversized Vehicle',
       'Fatigued/Drowsy', 'Alcohol Involvement',
       'Outside Car Distraction', 'Passenger Distraction',
       'Tinted Windows', 'Illnes', 'Lane Marking Improper/Inadequate',
       'Animals Action', 'Traffic Control Device Improper/Non-Working',
       'Accelerator Defective', 'Tire Failure/Inadequate',
       'Brakes Defective', 'Fell Asleep', 'Drive

contributing_factor_vehicle_2 had the same problem with 'Illnes' being missspelled, so we will fix it now.

In [24]:
datanyc['contributing_factor_vehicle_2'].replace('Illnes', 'Illness', inplace=True)

The crashes vehicle_type_code columns had a few similare problems, which we will clean now.

In [27]:
datanyc['vehicle_type_code_1'].replace('TAXI', 'Taxi', inplace=True)
datanyc['vehicle_type_code_2'].replace('TAXI', 'Taxi', inplace=True)
datanyc['vehicle_type_code_3'].replace('TAXI', 'Taxi', inplace=True)
datanyc['vehicle_type_code_4'].replace('TAXI', 'Taxi', inplace=True)
datanyc['vehicle_type_code_5'].replace('TAXI', 'Taxi', inplace=True)
datanyc['vehicle_type_code_1'].replace('PICK-UP TRUCK', 'Pick-up Truck', inplace=True)
datanyc['vehicle_type_code_2'].replace('PICK-UP TRUCK', 'Pick-up Truck', inplace=True)
datanyc['vehicle_type_code_3'].replace('PICK-UP TRUCK', 'Pick-up Truck', inplace=True)
datanyc['vehicle_type_code_4'].replace('PICK-UP TRUCK', 'Pick-up Truck', inplace=True)
datanyc['vehicle_type_code_5'].replace('PICK-UP TRUCK', 'Pick-up Truck', inplace=True)
datanyc['vehicle_type_code_1'].replace('AMBULANCE', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_2'].replace('AMBULANCE', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_3'].replace('AMBULANCE', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_4'].replace('AMBULANCE', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_5'].replace('AMBULANCE', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('SPORT UTILITY / STATION WAGON', 'Station Wagon/Sport Utility Vehicle', inplace=True)
datanyc['vehicle_type_code_2'].replace('SPORT UTILITY / STATION WAGON', 'Station Wagon/Sport Utility Vehicle', inplace=True)
datanyc['vehicle_type_code_3'].replace('SPORT UTILITY / STATION WAGON', 'Station Wagon/Sport Utility Vehicle', inplace=True)
datanyc['vehicle_type_code_4'].replace('SPORT UTILITY / STATION WAGON', 'Station Wagon/Sport Utility Vehicle', inplace=True)
datanyc['vehicle_type_code_5'].replace('SPORT UTILITY / STATION WAGON', 'Station Wagon/Sport Utility Vehicle', inplace=True)

In [28]:
datanyc['vehicle_type_code_1'].value_counts()

PASSENGER VEHICLE                      715236
Station Wagon/Sport Utility Vehicle    435791
Sedan                                  151360
Taxi                                    66305
Pick-up Truck                           33237
VAN                                     26540
OTHER                                   23982
UNKNOWN                                 19929
SMALL COM VEH(4 TIRES)                  14559
LARGE COM VEH(6 OR MORE TIRES)          14527
BUS                                     14057
LIVERY VEHICLE                          10481
Box Truck                                7965
Bus                                      6555
MOTORCYCLE                               6536
BICYCLE                                  5568
Bike                                     3916
Tractor Truck Diesel                     3496
Ambulance                                3339
Van                                      2708
TK                                       2485
BU                                

Next, we will focus just on vehicle_type_code_1. 

In [29]:
datanyc['vehicle_type_code_1'].replace('Other', 'Unknown', inplace=True)
datanyc['vehicle_type_code_1'].replace('BUS', 'Bus', inplace=True)
datanyc['vehicle_type_code_1'].replace('Box T', 'Box Truck', inplace=True)
datanyc['vehicle_type_code_1'].replace('GARBA', 'Garbage or Refuse', inplace=True)
datanyc['vehicle_type_code_1'].replace('CAB', 'Cab', inplace=True)
datanyc['vehicle_type_code_1'].replace('fdny', 'FIRE TRUCK', inplace=True)
datanyc['vehicle_type_code_1'].unique()

array(['Sedan', 'Bus', 'Ambulance', 'Station Wagon/Sport Utility Vehicle',
       'Pick-up Truck', 'Tractor Truck Diesel', 'Taxi', 'Bike',
       'Box Truck', 'Van', nan, 'Fire', 'Motorcycle',
       'Tractor Truck Gasoline', 'Convertible', 'E-Bik', 'Motorscooter',
       'prks', 'Flat Rack', 'PASSENGER VEHICLE', 'Dump', 'Flat Bed',
       'POWER', 'Garbage or Refuse', 'E-Sco', 'UTIL', 'Motorbike',
       'Trail', 'Tow Truck / Wrecker', 'Carry All', 'Ambul', 'tk', 'PK',
       'Multi-Wheeled Vehicle', 'LARGE COM VEH(6 OR MORE TIRES)',
       'Lift Boom', 'COMME', 'Moped', 'Tanker', 'Beverage Truck', 'MINI',
       'TRAIL', 'fire', 'CONT', 'Chassis Cab', 'Lunch Wagon',
       'Stake or Rack', 'Van Camper', 'REFG', 'LIMO', 'van', 'forkl',
       'School Bus', 'Motorized Home', 'FIRE', 'AMBUL', 'BoxTr',
       'Refrigerated Van', 'Concrete Mixer', '3-Door', 'PICK-', 'E BIK',
       'van t', 'DELIV', 'WINEB', 'Pickup with mounted Camper', 'Parce',
       'TRACT', 'HI TA', 'FDNY', '35 FT', 

In [30]:
datanyc['vehicle_type_code_1'].replace('VAN T', 'Van', inplace=True)
datanyc['vehicle_type_code_1'].replace('van', 'Van', inplace=True)
datanyc['vehicle_type_code_1'].replace('VAN/T', 'Van', inplace=True)
datanyc['vehicle_type_code_1'].replace('van t', 'Van', inplace=True)
datanyc['vehicle_type_code_1'].replace('van', 'Van', inplace=True)
datanyc['vehicle_type_code_1'].replace('VAN', 'Van', inplace=True)
datanyc['vehicle_type_code_1'].replace('AMBU', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('Ambul', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('AMB', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('Ambu', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('AMBUL', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('ambul', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('Ambu', 'Ambulance', inplace=True)
datanyc['vehicle_type_code_1'].replace('Fire', 'FIRE TRUCK', inplace=True)
datanyc['vehicle_type_code_1'].replace('fire', 'FIRE TRUCK', inplace=True)
datanyc['vehicle_type_code_1'].replace('FIRE', 'FIRE TRUCK', inplace=True)
datanyc['vehicle_type_code_1'].replace('FIRET', 'FIRE TRUCK', inplace=True)
datanyc['vehicle_type_code_1'].replace('FDNY', 'FIRE TRUCK', inplace=True)
datanyc['vehicle_type_code_1'].unique()

array(['Sedan', 'Bus', 'Ambulance', 'Station Wagon/Sport Utility Vehicle',
       'Pick-up Truck', 'Tractor Truck Diesel', 'Taxi', 'Bike',
       'Box Truck', 'Van', nan, 'FIRE TRUCK', 'Motorcycle',
       'Tractor Truck Gasoline', 'Convertible', 'E-Bik', 'Motorscooter',
       'prks', 'Flat Rack', 'PASSENGER VEHICLE', 'Dump', 'Flat Bed',
       'POWER', 'Garbage or Refuse', 'E-Sco', 'UTIL', 'Motorbike',
       'Trail', 'Tow Truck / Wrecker', 'Carry All', 'tk', 'PK',
       'Multi-Wheeled Vehicle', 'LARGE COM VEH(6 OR MORE TIRES)',
       'Lift Boom', 'COMME', 'Moped', 'Tanker', 'Beverage Truck', 'MINI',
       'TRAIL', 'CONT', 'Chassis Cab', 'Lunch Wagon', 'Stake or Rack',
       'Van Camper', 'REFG', 'LIMO', 'forkl', 'School Bus',
       'Motorized Home', 'BoxTr', 'Refrigerated Van', 'Concrete Mixer',
       '3-Door', 'PICK-', 'E BIK', 'DELIV', 'WINEB',
       'Pickup with mounted Camper', 'Parce', 'TRACT', 'HI TA', '35 FT',
       'MOPED', 'Bulk Agriculture', 'CITY', 'TRLR', 'Pedica

Even in one column you can see how much variation there is. We would suggest that whoever created this dataset turns this from a "fill in the blank" text field to a select field froma predetermined loist, to get better data fidelity, and we hope by this point, you can see why.

Next, we will rename some of the columns to make things easier while analyzing the data.

In [45]:
clean_nyc.rename(columns={'number_of_persons_injured' : 'persons_injured',
                        'number_of_persons_killed' : 'persons_killed',
                        'number_of_pedestrians_injured' : 'pedestrians_injured',
                        'number_of_pedestrians_killed' : 'pedestrians_killed',
                        'number_of_cyclist_injured' : 'cyclist_injured',
                        'number_of_cyclist_killed' : 'cyclist_killed',
                        'number_of_motorist_injured'  : 'motorist_injured',
                        'number_of_motorist_killed' : 'motorist_killed'},inplace=True)
clean_nyc.head()

date                time    borough zip_code   latitude  longitude  \
0 2019-08-08 2019-11-04 20:27:00  MANHATTAN  10033.0  40.848118 -73.930890   
1 2019-08-30 2019-11-04 19:00:00        NaN      nan  40.743214 -73.951350   
2 2019-08-18 2019-11-04 17:56:00        NaN      nan  40.839725 -73.783700   
3 2019-09-02 2019-11-04 01:00:00   BROOKLYN  11208.0  40.670254 -73.857280   
4 2019-08-08 2019-11-04 14:30:00        NaN      nan  40.728603 -74.005325   

   persons_injured  persons_killed  pedestrians_injured  pedestrians_killed  \
0              0.0             0.0                    0                   0   
1              0.0             0.0                    0                   0   
2              1.0             0.0                    0                   0   
3              0.0             0.0                    0                   0   
4              0.0             0.0                    0                   0   

   cyclist_injured  cyclist_killed  motorist_injured  motorist_killed  \
0                0               0                 0                0   
1                0               0                 0                0   
2                0               0                 1                0   
3                0               0                 0                0   
4                0               0                 0                0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  Driver Inattention/Distraction  Driver Inattention/Distraction   
1   Failure to Yield Right-of-Way                     Unspecified   
2                     Unspecified                     Unspecified   
3   Failure to Yield Right-of-Way                     Unspecified   
4  Driver Inattention/Distraction                     Unspecified   

  vehicle_type_code1                   vehicle_type_code2  hour  
0              Sedan  Station Wagon/Sport Utility Vehicle    20  
1              Sedan                                Sedan    19  
2              Sedan  Station Wagon/Sport Utility Vehicle    17  
3              Sedan                                  NaN     1  
4                Bus                            Box Truck    14

Great! We can see that 'persons_injured' and 'persons_killed' column values are float. Let's change them from float to integer. To do so, we will first get rid of any missing values in the 'persons_injured' and 'persons_killed' columns.

In [46]:
clean_nyc.dropna(subset = ['persons_injured'], how='all', inplace=True)
clean_nyc.dropna(subset = ['persons_killed'], how='all', inplace=True)
clean_nyc['persons_injured'] = clean_nyc.persons_injured.astype(int)
clean_nyc['persons_killed'] = clean_nyc.persons_killed.astype(int)
clean_nyc.head()

date                time    borough zip_code   latitude  longitude  \
0 2019-08-08 2019-11-04 20:27:00  MANHATTAN  10033.0  40.848118 -73.930890   
1 2019-08-30 2019-11-04 19:00:00        NaN      nan  40.743214 -73.951350   
2 2019-08-18 2019-11-04 17:56:00        NaN      nan  40.839725 -73.783700   
3 2019-09-02 2019-11-04 01:00:00   BROOKLYN  11208.0  40.670254 -73.857280   
4 2019-08-08 2019-11-04 14:30:00        NaN      nan  40.728603 -74.005325   

   persons_injured  persons_killed  pedestrians_injured  pedestrians_killed  \
0                0               0                    0                   0   
1                0               0                    0                   0   
2                1               0                    0                   0   
3                0               0                    0                   0   
4                0               0                    0                   0   

   cyclist_injured  cyclist_killed  motorist_injured  motorist_killed  \
0                0               0                 0                0   
1                0               0                 0                0   
2                0               0                 1                0   
3                0               0                 0                0   
4                0               0                 0                0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  Driver Inattention/Distraction  Driver Inattention/Distraction   
1   Failure to Yield Right-of-Way                     Unspecified   
2                     Unspecified                     Unspecified   
3   Failure to Yield Right-of-Way                     Unspecified   
4  Driver Inattention/Distraction                     Unspecified   

  vehicle_type_code1                   vehicle_type_code2  hour  
0              Sedan  Station Wagon/Sport Utility Vehicle    20  
1              Sedan                                Sedan    19  
2              Sedan  Station Wagon/Sport Utility Vehicle    17  
3              Sedan                                  NaN     1  
4                Bus                            Box Truck    14

Next, we want to change all my values to lower case letters.

In [47]:
clean_nyc1 = clean_nyc.applymap(lambda s:s.lower() if type(s) == str else s)
clean_nyc1.head()

date                time    borough zip_code   latitude  longitude  \
0 2019-08-08 2019-11-04 20:27:00  manhattan  10033.0  40.848118 -73.930890   
1 2019-08-30 2019-11-04 19:00:00        NaN      nan  40.743214 -73.951350   
2 2019-08-18 2019-11-04 17:56:00        NaN      nan  40.839725 -73.783700   
3 2019-09-02 2019-11-04 01:00:00   brooklyn  11208.0  40.670254 -73.857280   
4 2019-08-08 2019-11-04 14:30:00        NaN      nan  40.728603 -74.005325   

   persons_injured  persons_killed  pedestrians_injured  pedestrians_killed  \
0                0               0                    0                   0   
1                0               0                    0                   0   
2                1               0                    0                   0   
3                0               0                    0                   0   
4                0               0                    0                   0   

   cyclist_injured  cyclist_killed  motorist_injured  motorist_killed  \
0                0               0                 0                0   
1                0               0                 0                0   
2                0               0                 1                0   
3                0               0                 0                0   
4                0               0                 0                0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  driver inattention/distraction  driver inattention/distraction   
1   failure to yield right-of-way                     unspecified   
2                     unspecified                     unspecified   
3   failure to yield right-of-way                     unspecified   
4  driver inattention/distraction                     unspecified   

  vehicle_type_code1                   vehicle_type_code2  hour  
0              sedan  station wagon/sport utility vehicle    20  
1              sedan                                sedan    19  
2              sedan  station wagon/sport utility vehicle    17  
3              sedan                                  NaN     1  
4                bus                            box truck    14

We replaced the missing values in the 'latitude' and 'longitude' columns with zeros, now we also want to do something with the missing values in other columns. Let's see what values we have in the 'contributing_factor_vehicle_1' and 'contributing_factor_vehicle_2' columns.

In [48]:
clean_nyc1['contributing_factor_vehicle_1'].value_counts(dropna=False)

unspecified                                              587941
driver inattention/distraction                           296057
failure to yield right-of-way                             90524
following too closely                                     78979
backing unsafely                                          61010
other vehicular                                           50753
fatigued/drowsy                                           46847
turning improperly                                        40834
passing or lane usage improper                            37277
passing too closely                                       34169
unsafe lane changing                                      30301
traffic control disregarded                               24341
driver inexperience                                       23429
lost consciousness                                        19617
pavement slippery                                         15788
prescription medication                 

We can combine the missing and unspecified values.

In [49]:
clean_nyc1['contributing_factor_vehicle_1'].fillna(value='Unspecified', inplace=True)
clean_nyc1['contributing_factor_vehicle_1'].value_counts(dropna=False)

unspecified                                              587941
driver inattention/distraction                           296057
failure to yield right-of-way                             90524
following too closely                                     78979
backing unsafely                                          61010
other vehicular                                           50753
fatigued/drowsy                                           46847
turning improperly                                        40834
passing or lane usage improper                            37277
passing too closely                                       34169
unsafe lane changing                                      30301
traffic control disregarded                               24341
driver inexperience                                       23429
lost consciousness                                        19617
pavement slippery                                         15788
prescription medication                 

Let's see the values for the 'contributing_factor_vehicle_2' column.

In [50]:
clean_nyc1['contributing_factor_vehicle_2'].value_counts(dropna=False)

unspecified                                              1161335
NaN                                                       214219
driver inattention/distraction                             72602
other vehicular                                            26624
failure to yield right-of-way                              13888
following too closely                                      13389
fatigued/drowsy                                            10823
passing or lane usage improper                              9291
turning improperly                                          7688
backing unsafely                                            6976
passing too closely                                         6514
driver inexperience                                         5440
traffic control disregarded                                 5273
lost consciousness                                          5223
unsafe lane changing                                        5222
pavement slippery        

We can do the same for this column as well.

In [51]:
clean_nyc1['contributing_factor_vehicle_2'].fillna(value='Unspecified', inplace=True)
clean_nyc1['contributing_factor_vehicle_1'].value_counts(dropna=False)

unspecified                                              587941
driver inattention/distraction                           296057
failure to yield right-of-way                             90524
following too closely                                     78979
backing unsafely                                          61010
other vehicular                                           50753
fatigued/drowsy                                           46847
turning improperly                                        40834
passing or lane usage improper                            37277
passing too closely                                       34169
unsafe lane changing                                      30301
traffic control disregarded                               24341
driver inexperience                                       23429
lost consciousness                                        19617
pavement slippery                                         15788
prescription medication                 

## Step 3 - 'Latitude' and 'longitude' Columns

The 'location' column is simply a concatenation of 'latitude' and 'longitude' columns.

In [32]:
clean_nyc[["latitude", "longitude", "location"]].head(30)

latitude  longitude                      location
0   40.848118 -73.930890   POINT (-73.93089 40.848118)
1   40.743214 -73.951350   POINT (-73.95135 40.743214)
2   40.839725 -73.783700    POINT (-73.7837 40.839725)
3   40.670254 -73.857280   POINT (-73.85728 40.670254)
4   40.728603 -74.005325  POINT (-74.005325 40.728603)
5   40.784943 -73.946594  POINT (-73.946594 40.784943)
6   40.733536 -73.870350   POINT (-73.87035 40.733536)
7   40.676900 -73.781580     POINT (-73.78158 40.6769)
8   40.689690 -73.992370    POINT (-73.99237 40.68969)
9   40.656610 -73.930910    POINT (-73.93091 40.65661)
10  40.602600 -74.014910     POINT (-74.01491 40.6026)
11  40.818172 -73.956490   POINT (-73.95649 40.818172)
12  40.745780 -73.991806   POINT (-73.991806 40.74578)
13  40.630478 -73.986100    POINT (-73.9861 40.630478)
14  40.615260 -73.904840    POINT (-73.90484 40.61526)
15  40.731277 -73.871574  POINT (-73.871574 40.731277)
16  40.738045 -73.996346  POINT (-73.996346 40.738045)
17  40.773530 -73.985340    POINT (-73.98534 40.77353)
18  40.875180 -73.867050    POINT (-73.86705 40.87518)
19  40.712532 -74.007675  POINT (-74.007675 40.712532)
20  40.634250 -74.010850    POINT (-74.01085 40.63425)
21  40.689297 -73.874060   POINT (-73.87406 40.689297)
22  40.747020 -74.004660    POINT (-74.00466 40.74702)
23  40.703957 -73.895700    POINT (-73.8957 40.703957)
24  40.813420 -73.937150    POINT (-73.93715 40.81342)
25        NaN        NaN                           NaN
26  40.801420 -73.934460    POINT (-73.93446 40.80142)
27  40.685947 -73.802240   POINT (-73.80224 40.685947)
28  40.729134 -73.951810   POINT (-73.95181 40.729134)
29  40.709690 -73.846030    POINT (-73.84603 40.70969)

Before we quickly do something, let's first see if it's true that all the location data follows the same pattern we see right now:

In [33]:
clean_nyc['location'].str.match('POINT \(-7\d\.\d+ \d{2}\.\d+\)', na=False).value_counts()

True     1399184
False     196679
Name: location, dtype: int64

196679 rows where that's not the case! More than we expected! Let's check them out!

In [34]:
clean_nyc[~clean_nyc['location'].str.match('POINT \(-7\d\.\d+ \d{2}\.\d+\)', na=False)]

date   time        borough zip_code  latitude  \
25       2019-08-04T00:00:00.000   4:01            NaN      nan       NaN   
36       2019-08-29T00:00:00.000  15:34            NaN      nan       NaN   
57       2019-08-10T00:00:00.000   3:45            NaN      nan       NaN   
62       2019-08-13T00:00:00.000   9:30            NaN      nan       NaN   
71       2019-08-24T00:00:00.000  23:50            NaN      nan       NaN   
88       2019-08-22T00:00:00.000  16:36            NaN      nan       NaN   
90       2019-08-16T00:00:00.000  11:20            NaN      nan       NaN   
96       2019-09-01T00:00:00.000  17:03            NaN      nan       NaN   
97       2019-08-23T00:00:00.000  18:47            NaN      nan       NaN   
98       2019-08-09T00:00:00.000  10:00            NaN      nan       NaN   
103      2019-08-02T00:00:00.000  15:00            NaN      nan       NaN   
116      2019-08-10T00:00:00.000  11:00            NaN      nan       NaN   
121      2019-08-08T00:00:00.000  12:28            NaN      nan       NaN   
129      2019-10-29T00:00:00.000  13:55            NaN      nan       NaN   
154      2019-08-28T00:00:00.000   0:13            NaN      nan       NaN   
155      2019-10-29T00:00:00.000  15:59            NaN      nan       NaN   
156      2019-08-30T00:00:00.000  17:30            NaN      nan       NaN   
159      2019-08-23T00:00:00.000   0:45            NaN      nan       NaN   
172      2019-08-22T00:00:00.000  20:00            NaN      nan       NaN   
177      2019-08-29T00:00:00.000  19:00            NaN      nan       NaN   
182      2019-10-29T00:00:00.000  16:40            NaN      nan       NaN   
192      2019-08-24T00:00:00.000  17:50         QUEENS  11697.0       NaN   
197      2019-08-27T00:00:00.000  13:15            NaN      nan       NaN   
199      2019-08-20T00:00:00.000  15:00            NaN      nan       NaN   
204      2019-08-30T00:00:00.000  14:50         QUEENS  11355.0       NaN   
210      2019-10-29T00:00:00.000  17:32            NaN      nan       NaN   
229      2019-09-03T00:00:00.000   9:10            NaN      nan       NaN   
230      2019-08-27T00:00:00.000  13:30            NaN      nan       NaN   
281      2019-08-06T00:00:00.000   9:45            NaN      nan       NaN   
287      2019-10-29T00:00:00.000  19:41            NaN      nan       NaN   
308      2019-08-06T00:00:00.000   9:08      MANHATTAN  10019.0       NaN   
309      2019-08-30T00:00:00.000  11:15            NaN      nan       NaN   
310      2019-08-12T00:00:00.000  12:45            NaN      nan       NaN   
314      2016-03-26T00:00:00.000  15:19      MANHATTAN  10033.0       NaN   
315      2019-10-29T00:00:00.000  20:06            NaN      nan       NaN   
347      2019-08-27T00:00:00.000  23:45            NaN      nan       NaN   
350      2019-08-08T00:00:00.000  18:00            NaN      nan       NaN   
362      2019-08-24T00:00:00.000  14:40            NaN      nan       NaN   
368      2019-10-29T00:00:00.000   2:30            NaN      nan       NaN   
379      2019-08-23T00:00:00.000  22:52            NaN      nan       NaN   
384      2019-09-01T00:00:00.000  14:30            NaN      nan       NaN   
397      2019-08-15T00:00:00.000  23:50            NaN      nan       NaN   
400      2019-08-16T00:00:00.000   0:48            NaN      nan       NaN   
405      2019-08-12T00:00:00.000  17:03       BROOKLYN  11239.0       NaN   
421      2019-10-29T00:00:00.000   7:05            NaN      nan       NaN   
445      2019-08-10T00:00:00.000  14:00          BRONX  10475.0       NaN   
447      2019-08-17T00:00:00.000  21:30            NaN      nan       NaN   
454      2019-08-25T00:00:00.000   4:28            NaN      nan       NaN   
495      2019-08-06T00:00:00.000   9:00            NaN      nan       NaN   
504      2019-08-14T00:00:00.000  15:10            NaN      nan       NaN   
513      2019-08-06T00:00:00.000  19:05            NaN      nan       NaN   
525      2019-09-02T00:0

Ahh those sweet missing values... As we've seen before, latitude, longitude	and location columns have 12% of their values missing each. While we're at it, let's make some changes with the help of fillna function.

In [35]:
clean_nyc['location'] = clean_nyc['location'].fillna(0)
clean_nyc['latitude'] = clean_nyc['latitude'].fillna(0)
clean_nyc['longitude'] = clean_nyc['longitude'].fillna(0)

In [36]:
clean_nyc[~clean_nyc['location'].str.match('POINT \(-7\d\.\d+ \d{2}\.\d+\)', na=False)]

date   time        borough zip_code  latitude  \
25       2019-08-04T00:00:00.000   4:01            NaN      nan       0.0   
36       2019-08-29T00:00:00.000  15:34            NaN      nan       0.0   
57       2019-08-10T00:00:00.000   3:45            NaN      nan       0.0   
62       2019-08-13T00:00:00.000   9:30            NaN      nan       0.0   
71       2019-08-24T00:00:00.000  23:50            NaN      nan       0.0   
88       2019-08-22T00:00:00.000  16:36            NaN      nan       0.0   
90       2019-08-16T00:00:00.000  11:20            NaN      nan       0.0   
96       2019-09-01T00:00:00.000  17:03            NaN      nan       0.0   
97       2019-08-23T00:00:00.000  18:47            NaN      nan       0.0   
98       2019-08-09T00:00:00.000  10:00            NaN      nan       0.0   
103      2019-08-02T00:00:00.000  15:00            NaN      nan       0.0   
116      2019-08-10T00:00:00.000  11:00            NaN      nan       0.0   
121      2019-08-08T00:00:00.000  12:28            NaN      nan       0.0   
129      2019-10-29T00:00:00.000  13:55            NaN      nan       0.0   
154      2019-08-28T00:00:00.000   0:13            NaN      nan       0.0   
155      2019-10-29T00:00:00.000  15:59            NaN      nan       0.0   
156      2019-08-30T00:00:00.000  17:30            NaN      nan       0.0   
159      2019-08-23T00:00:00.000   0:45            NaN      nan       0.0   
172      2019-08-22T00:00:00.000  20:00            NaN      nan       0.0   
177      2019-08-29T00:00:00.000  19:00            NaN      nan       0.0   
182      2019-10-29T00:00:00.000  16:40            NaN      nan       0.0   
192      2019-08-24T00:00:00.000  17:50         QUEENS  11697.0       0.0   
197      2019-08-27T00:00:00.000  13:15            NaN      nan       0.0   
199      2019-08-20T00:00:00.000  15:00            NaN      nan       0.0   
204      2019-08-30T00:00:00.000  14:50         QUEENS  11355.0       0.0   
210      2019-10-29T00:00:00.000  17:32            NaN      nan       0.0   
229      2019-09-03T00:00:00.000   9:10            NaN      nan       0.0   
230      2019-08-27T00:00:00.000  13:30            NaN      nan       0.0   
281      2019-08-06T00:00:00.000   9:45            NaN      nan       0.0   
287      2019-10-29T00:00:00.000  19:41            NaN      nan       0.0   
308      2019-08-06T00:00:00.000   9:08      MANHATTAN  10019.0       0.0   
309      2019-08-30T00:00:00.000  11:15            NaN      nan       0.0   
310      2019-08-12T00:00:00.000  12:45            NaN      nan       0.0   
314      2016-03-26T00:00:00.000  15:19      MANHATTAN  10033.0       0.0   
315      2019-10-29T00:00:00.000  20:06            NaN      nan       0.0   
347      2019-08-27T00:00:00.000  23:45            NaN      nan       0.0   
350      2019-08-08T00:00:00.000  18:00            NaN      nan       0.0   
362      2019-08-24T00:00:00.000  14:40            NaN      nan       0.0   
368      2019-10-29T00:00:00.000   2:30            NaN      nan       0.0   
379      2019-08-23T00:00:00.000  22:52            NaN      nan       0.0   
384      2019-09-01T00:00:00.000  14:30            NaN      nan       0.0   
397      2019-08-15T00:00:00.000  23:50            NaN      nan       0.0   
400      2019-08-16T00:00:00.000   0:48            NaN      nan       0.0   
405      2019-08-12T00:00:00.000  17:03       BROOKLYN  11239.0       0.0   
421      2019-10-29T00:00:00.000   7:05            NaN      nan       0.0   
445      2019-08-10T00:00:00.000  14:00          BRONX  10475.0       0.0   
447      2019-08-17T00:00:00.000  21:30            NaN      nan       0.0   
454      2019-08-25T00:00:00.000   4:28            NaN      nan       0.0   
495      2019-08-06T00:00:00.000   9:00            NaN      nan       0.0   
504      2019-08-14T00:00:00.000  15:10            NaN      nan       0.0   
513      2019-08-06T00:00:00.000  19:05            NaN      nan       0.0   
525      2019-09-02T00:0

Now all NaN's in the latitude and longitude columns are replaced with 0, let's change latitude, let's drop the location column.

In [38]:
clean_nyc.drop(columns="location", inplace = True)

KeyError: "['location'] not found in axis"

And let's peek at our data.

In [39]:
clean_nyc.head()

date   time    borough zip_code   latitude  longitude  \
0  2019-08-08T00:00:00.000  20:27  MANHATTAN  10033.0  40.848118 -73.930890   
1  2019-08-30T00:00:00.000  19:00        NaN      nan  40.743214 -73.951350   
2  2019-08-18T00:00:00.000  17:56        NaN      nan  40.839725 -73.783700   
3  2019-09-02T00:00:00.000   1:00   BROOKLYN  11208.0  40.670254 -73.857280   
4  2019-08-08T00:00:00.000  14:30        NaN      nan  40.728603 -74.005325   

   number_of_persons_injured  number_of_persons_killed  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        1.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   number_of_pedestrians_injured  number_of_pedestrians_killed  \
0                              0                             0   
1                              0                             0   
2                              0                             0   
3                              0                             0   
4                              0                             0   

   number_of_cyclist_injured  number_of_cyclist_killed  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   number_of_motorist_injured  number_of_motorist_killed  \
0                           0                          0   
1                           0                          0   
2                           1                          0   
3                           0                          0   
4                           0                          0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  Driver Inattention/Distraction  Driver Inattention/Distraction   
1   Failure to Yield Right-of-Way                     Unspecified   
2                     Unspecified                     Unspecified   
3   Failure to Yield Right-of-Way                     Unspecified   
4  Driver Inattention/Distraction                     Unspecified   

  vehicle_type_code1                   vehicle_type_code2  
0              Sedan  Station Wagon/Sport Utility Vehicle  
1              Sedan                                Sedan  
2              Sedan  Station Wagon/Sport Utility Vehicle  
3              Sedan                                  NaN  
4                Bus                            Box Truck

In [40]:
clean_nyc['latitude'] = clean_nyc['latitude'].fillna(0)
clean_nyc['longitude'] = clean_nyc['longitude'].fillna(0)
clean_nyc.head()

date   time    borough zip_code   latitude  longitude  \
0  2019-08-08T00:00:00.000  20:27  MANHATTAN  10033.0  40.848118 -73.930890   
1  2019-08-30T00:00:00.000  19:00        NaN      nan  40.743214 -73.951350   
2  2019-08-18T00:00:00.000  17:56        NaN      nan  40.839725 -73.783700   
3  2019-09-02T00:00:00.000   1:00   BROOKLYN  11208.0  40.670254 -73.857280   
4  2019-08-08T00:00:00.000  14:30        NaN      nan  40.728603 -74.005325   

   number_of_persons_injured  number_of_persons_killed  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        1.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   number_of_pedestrians_injured  number_of_pedestrians_killed  \
0                              0                             0   
1                              0                             0   
2                              0                             0   
3                              0                             0   
4                              0                             0   

   number_of_cyclist_injured  number_of_cyclist_killed  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   number_of_motorist_injured  number_of_motorist_killed  \
0                           0                          0   
1                           0                          0   
2                           1                          0   
3                           0                          0   
4                           0                          0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  Driver Inattention/Distraction  Driver Inattention/Distraction   
1   Failure to Yield Right-of-Way                     Unspecified   
2                     Unspecified                     Unspecified   
3   Failure to Yield Right-of-Way                     Unspecified   
4  Driver Inattention/Distraction                     Unspecified   

  vehicle_type_code1                   vehicle_type_code2  
0              Sedan  Station Wagon/Sport Utility Vehicle  
1              Sedan                                Sedan  
2              Sedan  Station Wagon/Sport Utility Vehicle  
3              Sedan                                  NaN  
4                Bus                            Box Truck

## Step 4: Date Format

Let's look at our dates to make sure they are all in the same format:

In [42]:
clean_nyc[['date', 'time']].head()

date   time
0  2019-08-08T00:00:00.000  20:27
1  2019-08-30T00:00:00.000  19:00
2  2019-08-18T00:00:00.000  17:56
3  2019-09-02T00:00:00.000   1:00
4  2019-08-08T00:00:00.000  14:30

The 'date' column definitely needs some fixing... We will transform the string timestamp for 'date' to a true datetime data type.

In [43]:
clean_nyc['date'] = pd.to_datetime(datanyc['date'])
clean_nyc.head()

date   time    borough zip_code   latitude  longitude  \
0 2019-08-08  20:27  MANHATTAN  10033.0  40.848118 -73.930890   
1 2019-08-30  19:00        NaN      nan  40.743214 -73.951350   
2 2019-08-18  17:56        NaN      nan  40.839725 -73.783700   
3 2019-09-02   1:00   BROOKLYN  11208.0  40.670254 -73.857280   
4 2019-08-08  14:30        NaN      nan  40.728603 -74.005325   

   number_of_persons_injured  number_of_persons_killed  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        1.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   number_of_pedestrians_injured  number_of_pedestrians_killed  \
0                              0                             0   
1                              0                             0   
2                              0                             0   
3                              0                             0   
4                              0                             0   

   number_of_cyclist_injured  number_of_cyclist_killed  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   number_of_motorist_injured  number_of_motorist_killed  \
0                           0                          0   
1                           0                          0   
2                           1                          0   
3                           0                          0   
4                           0                          0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  Driver Inattention/Distraction  Driver Inattention/Distraction   
1   Failure to Yield Right-of-Way                     Unspecified   
2                     Unspecified                     Unspecified   
3   Failure to Yield Right-of-Way                     Unspecified   
4  Driver Inattention/Distraction                     Unspecified   

  vehicle_type_code1                   vehicle_type_code2  
0              Sedan  Station Wagon/Sport Utility Vehicle  
1              Sedan                                Sedan  
2              Sedan  Station Wagon/Sport Utility Vehicle  
3              Sedan                                  NaN  
4                Bus                            Box Truck

We also want to create a new column which will carry the values for hours only. We think that can be helpful when grouping the times and visualizing the data. We will create a new column called 'hour' in which we will only have the hours instead of hours and minutes.

In [44]:
clean_nyc['time'] = pd.to_datetime(clean_nyc.time)
clean_nyc['hour'] = clean_nyc['time'].dt.hour
clean_nyc

date                time        borough zip_code   latitude  \
0       2019-08-08 2019-11-04 20:27:00      MANHATTAN  10033.0  40.848118   
1       2019-08-30 2019-11-04 19:00:00            NaN      nan  40.743214   
2       2019-08-18 2019-11-04 17:56:00            NaN      nan  40.839725   
3       2019-09-02 2019-11-04 01:00:00       BROOKLYN  11208.0  40.670254   
4       2019-08-08 2019-11-04 14:30:00            NaN      nan  40.728603   
5       2019-09-01 2019-11-04 12:55:00            NaN      nan  40.784943   
6       2019-09-03 2019-11-04 17:15:00         QUEENS  11373.0  40.733536   
7       2019-08-14 2019-11-04 19:04:00         QUEENS  11434.0  40.676900   
8       2019-08-05 2019-11-04 17:25:00       BROOKLYN  11201.0  40.689690   
9       2019-08-24 2019-11-04 01:39:00       BROOKLYN  11203.0  40.656610   
10      2019-08-23 2019-11-04 15:34:00            NaN      nan  40.602600   
11      2019-08-20 2019-11-04 13:15:00      MANHATTAN  10027.0  40.818172   
12      2019-08-23 2019-11-04 11:35:00      MANHATTAN  10001.0  40.745780   
13      2019-08-07 2019-11-04 08:57:00       BROOKLYN  11204.0  40.630478   
14      2019-08-15 2019-11-04 00:30:00       BROOKLYN  11234.0  40.615260   
15      2019-08-12 2019-11-04 01:40:00         QUEENS  11373.0  40.731277   
16      2019-08-10 2019-11-04 21:07:00      MANHATTAN  10011.0  40.738045   
17      2019-08-30 2019-11-04 01:25:00            NaN      nan  40.773530   
18      2019-08-19 2019-11-04 16:20:00          BRONX  10467.0  40.875180   
19      2019-08-28 2019-11-04 10:00:00      MANHATTAN  10007.0  40.712532   
20      2019-08-09 2019-11-04 04:33:00       BROOKLYN  11220.0  40.634250   
21      2019-08-18 2019-11-04 15:10:00       BROOKLYN  11208.0  40.689297   
22      2019-08-24 2019-11-04 22:29:00      MANHATTAN  10011.0  40.747020   
23      2019-08-31 2019-11-04 11:10:00         QUEENS  11385.0  40.703957   
24      2019-08-31 2019-11-04 23:20:00      MANHATTAN  10037.0  40.813420   
25      2019-08-04 2019-11-04 04:01:00            NaN      nan   0.000000   
26      2019-08-05 2019-11-04 15:00:00            NaN      nan  40.801420   
27      2019-08-24 2019-11-04 04:28:00            NaN      nan  40.685947   
28      2019-08-23 2019-11-04 12:11:00       BROOKLYN  11222.0  40.729134   
29      2019-09-01 2019-11-04 17:30:00            NaN      nan  40.709690   
30      2019-08-15 2019-11-04 11:45:00            NaN      nan  40.759575   
31      2019-08-29 2019-11-04 20:40:00         QUEENS  11004.0  40.737743   
32      2019-08-20 2019-11-04 11:15:00      MANHATTAN  10035.0  40.803700   
33      2019-08-07 2019-11-04 21:26:00          BRONX  10474.0  40.806740   
34      2019-08-15 2019-11-04 12:40:00  STATEN ISLAND  10309.0  40.516457   
35      2019-08-17 2019-11-04 18:30:00       BROOKLYN  11234.0  40.618877   
36      2019-08-29 2019-11-04 15:34:00            NaN      nan   0.000000   
37      2019-08-10 2019-11-04 05:15:00            NaN      nan  40.674953   
38      2019-08-05 2019-11-04 16:09:00       BROOKLYN  11207.0  40.670970   
39      2019-08-23 2019-11-04 08:30:00          BRONX  10454.0  40.807365   
40      2019-08-09 2019-11-04 13:25:00       BROOKLYN  11210.0  40.638428   
41      2019-08-27 2019-11-04 18:40:00      MANHATTAN  10016.0  40.748850   
42      2019-08-14 2019-11-04 04:47:00         QUEENS  11412.0  40.691240   
43      2019-08-26 2019-11-04 13:40:00      MANHATTAN  10027.0  40.812725   
44      2019-08-05 2019-11-04 12:00:00       BROOKLYN  11208.0  40.679700   
45      2019-08-31 2019-11-04 01:30:00         QUEENS  11373.0  40.746098   
46      2019-08-18 2019-11-04 18:11:00       BROOKLYN  11207.0  40.678160   
47      2019-08-31 2019-11-04 16:38:00            NaN      nan  40.747242   
48      2019-08-26 2019-11-04 14:00:00      MANHATTAN  10065.0  40.764160   
49      2019-08-29 2019-11-04 11:09:00       BROOKLYN  11210.0  40.621857   
50      2019-08-17 2019-11-04 13:19:00      MANHATTAN  10065.0  40.761787   
51      20

## Step 5: Making a Seasons Variable

We are interested in adding a variable that gives the season in which something occurred.

In [53]:
clean_nyc1['date'] = pd.to_datetime(datanyc.date)
clean_nyc1['date'].dt.month

0          8
1          8
2          8
3          9
4          8
5          9
6          9
7          8
8          8
9          8
10         8
11         8
12         8
13         8
14         8
15         8
16         8
17         8
18         8
19         8
20         8
21         8
22         8
23         8
24         8
25         8
26         8
27         8
28         8
29         9
30         8
31         8
32         8
33         8
34         8
35         8
36         8
37         8
38         8
39         8
40         8
41         8
42         8
43         8
44         8
45         8
46         8
47         8
48         8
49         8
50         8
51         8
52         8
53         8
54         8
55         8
56         8
57         8
58         8
59         8
60         8
61         8
62         8
63         8
64         8
65         8
66         8
67         8
68         8
69         8
70         8
71         8
72         8
73         8
74         8
75         8
76         8

In [54]:
def season(date):
    if date.month in ([3, 4, 5]):
        val = 'Spring'
    elif date.month in ([6, 7, 8]):
        val = 'Summer'
    elif date.month in ([9, 10, 11]):
        val = 'Autumn'
    elif date.month in ([12, 1, 2]):
        val = 'Winter'
    else:
        val = "Unspecified"
    return val

clean_nyc1['season'] = clean_nyc1['date'].apply(season)

In [55]:
clean_nyc1['season'].value_counts()

Summer    431934
Autumn    422055
Spring    384121
Winter    357717
Name: season, dtype: int64

## Step 6: General Analysis

Let's see, for example, how many persons, cyclist, pedestrian and motorist killed in 3000000 collisions.

In [74]:
personskillednum = clean_nyc1['persons_killed'] > 0
personskillednum.sum()

1791

In [75]:
cyclistkillednum = clean_nyc1['cyclist_killed'] > 0
cyclistkillednum.sum()

133

In [76]:
motoristkillednum = clean_nyc1['motorist_killed'] > 0
motoristkillednum.sum()

674

In [77]:
pedestrianskillednum = clean_nyc1['pedestrians_killed'] > 0
pedestrianskillednum.sum()

991

It looks like being a pedestrian is not very safe in NYC. What about injured people? 

In [79]:
personsinjurednum = clean_nyc1['persons_injured'] > 0
personsinjurednum.sum()

306920

In [80]:
cyclistinjurednum = clean_nyc1['cyclist_injured'] > 0
cyclistinjurednum.sum()

32978

In [81]:
motoristinjurednum = clean_nyc1['motorist_injured'] > 0
motoristinjurednum.sum()

198428

In [82]:
pedestriansinjurednum = clean_nyc1['pedestrians_injured'] > 0
pedestriansinjurednum.sum()

77172

Interestingly, as far as injuries go, being a motorist is worse than being a pedestrian. So maybe we can say that there are many motorcycle collisions that result in injuries but not death?  

Let's see which boroughs have a high number of collisions.

In [83]:
datanyc['borough'].value_counts()

BROOKLYN         345363
QUEENS           296100
MANHATTAN        267536
BRONX            154889
STATEN ISLAND     48352
Name: borough, dtype: int64

A lot of collisions in Brooklyn! 

Let's see the common reasons behind collisions.

In [84]:
datanyc.contributing_factor_vehicle_1.value_counts()

Unspecified                                              587958
Driver Inattention/Distraction                           296063
Failure to Yield Right-of-Way                             90524
Following Too Closely                                     78979
Backing Unsafely                                          61010
Other Vehicular                                           50754
Fatigued/Drowsy                                           46847
Turning Improperly                                        40835
Passing or Lane Usage Improper                            37277
Passing Too Closely                                       34169
Unsafe Lane Changing                                      30302
Traffic Control Disregarded                               24341
Driver Inexperience                                       23430
Lost Consciousness                                        19618
Pavement Slippery                                         15788
Prescription Medication                 

Driver inattention/distraction and following too closely are the main causes of collisions. Do we need more proof why we shouldn't "send a quick text" while driving!?

What about the vehicle types? We will first make a copy of data -- a whole new data frame named contributing_factor to see the common vehicle types involved in driver inattention/distraction caused collisions.

In [86]:
contributing_factor = datanyc[datanyc['contributing_factor_vehicle_1'] == "Driver Inattention/Distraction"].copy()
contributing_factor.head()

date   time    borough zip_code   latitude  longitude  \
0   2019-08-08T00:00:00.000  20:27  MANHATTAN    10033  40.848118 -73.930890   
4   2019-08-08T00:00:00.000  14:30        NaN      NaN  40.728603 -74.005325   
8   2019-08-05T00:00:00.000  17:25   BROOKLYN    11201  40.689690 -73.992370   
10  2019-08-23T00:00:00.000  15:34        NaN      NaN  40.602600 -74.014910   
12  2019-08-23T00:00:00.000  11:35  MANHATTAN    10001  40.745780 -73.991806   

                        location                    on_street_name  \
0    POINT (-73.93089 40.848118)  AMSTERDAM AVENUE                   
4   POINT (-74.005325 40.728603)  WEST HOUSTON STREET                
8     POINT (-73.99237 40.68969)  ATLANTIC AVENUE                    
10     POINT (-74.01491 40.6026)  BELT PARKWAY                       
12   POINT (-73.991806 40.74578)                               NaN   

    off_street_name                         cross_street_name  \
0   WEST 181 STREET                                       NaN   
4               NaN                                       NaN   
8      COURT STREET                                       NaN   
10              NaN                                       NaN   
12              NaN  115       WEST 27 STREET                   

    number_of_persons_injured  number_of_persons_killed  \
0                         0.0                       0.0   
4                         0.0                       0.0   
8                         3.0                       0.0   
10                        0.0                       0.0   
12                        0.0                       0.0   

    number_of_pedestrians_injured  number_of_pedestrians_killed  \
0                               0                             0   
4                               0                             0   
8                               0                             0   
10                              0                             0   
12                              0                             0   

    number_of_cyclist_injured  number_of_cyclist_killed  \
0                           0                         0   
4                           0                         0   
8                           0                         0   
10                          0                         0   
12                          0                         0   

    number_of_motorist_injured  number_of_motorist_killed  \
0                            0                          0   
4                            0                          0   
8                            3                          0   
10                           0                          0   
12                           0                          0   

     contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0   Driver Inattention/Distraction  Driver Inattention/Distraction   
4   Driver Inattention/Distraction                     Unspecified   
8   Driver Inattention/Distraction                     Unspecified   
10  Driver Inattention/Distraction                     Unspecified   
12  Driver Inattention/Distraction                     Unspecified   

   contributing_factor_vehicle_3 contributing_factor_vehicle_4  \
0                            NaN                           NaN   
4                            NaN                           NaN   
8                    Unspecified                   Unspecified   
10                           NaN                           NaN   
12                           NaN                           NaN   

   contributing_factor_vehicle_5  unique_key vehicle_type_code_1  \
0                            NaN     4185696               Sedan   
4                            NaN     4185430                 Bus   
8                            NaN     4187679                 SUV   
10                           NaN     4193724                 SUV   
12                           NaN     4193584                 SUV   

                    veh

In [88]:
contributing_factor['vehicle_type_code_1'].value_counts()

PASSENGER VEHICLE                 118553
SUV                                88091
Sedan                              38247
Taxi                               14232
Pick-up Truck                       6830
Van                                 4457
Bus                                 3308
OTHER                               2442
SMALL COM VEH(4 TIRES)              2124
LARGE COM VEH(6 OR MORE TIRES)      2048
UNKNOWN                             1994
Box Truck                           1980
LIVERY VEHICLE                      1705
BICYCLE                             1321
Bike                                 962
MOTORCYCLE                           823
Tractor Truck Diesel                 702
TK                                   577
Ambulance                            541
BU                                   449
Motorcycle                           376
Convertible                          346
Dump                                 282
DS                                   202
4 dr sedan      

Passenger vehicles take the lead in terms of vehicles most commonly involced in driver inattention/distraction caused collisions. They are followed by sport utilities/station wagons. 

## Cleaning Conclusion

We have a more precise set of data to work with in clean_nyc. Let's take a peek at it.

In [66]:
clean_nyc1.shape

(1595827, 20)

In [67]:
datanyc.shape

(1595863, 29)

We've reduced our data by 9 columns and around 36 rows. Let's take a general look at our data as well as the 'info'.

In [68]:
clean_nyc1.head()

date                time    borough zip_code   latitude  longitude  \
0 2019-08-08 2019-11-04 20:27:00  manhattan  10033.0  40.848118 -73.930890   
1 2019-08-30 2019-11-04 19:00:00        NaN      nan  40.743214 -73.951350   
2 2019-08-18 2019-11-04 17:56:00        NaN      nan  40.839725 -73.783700   
3 2019-09-02 2019-11-04 01:00:00   brooklyn  11208.0  40.670254 -73.857280   
4 2019-08-08 2019-11-04 14:30:00        NaN      nan  40.728603 -74.005325   

   persons_injured  persons_killed  pedestrians_injured  pedestrians_killed  \
0                0               0                    0                   0   
1                0               0                    0                   0   
2                1               0                    0                   0   
3                0               0                    0                   0   
4                0               0                    0                   0   

   cyclist_injured  cyclist_killed  motorist_injured  motorist_killed  \
0                0               0                 0                0   
1                0               0                 0                0   
2                0               0                 1                0   
3                0               0                 0                0   
4                0               0                 0                0   

    contributing_factor_vehicle_1   contributing_factor_vehicle_2  \
0  driver inattention/distraction  driver inattention/distraction   
1   failure to yield right-of-way                     unspecified   
2                     unspecified                     unspecified   
3   failure to yield right-of-way                     unspecified   
4  driver inattention/distraction                     unspecified   

  vehicle_type_code1                   vehicle_type_code2  hour  season  
0              sedan  station wagon/sport utility vehicle    20  Summer  
1              sedan                                sedan    19  Summer  
2              sedan  station wagon/sport utility vehicle    17  Summer  
3              sedan                                  NaN     1  Autumn  
4                bus                            box truck    14  Summer

In [69]:
clean_nyc1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1595827 entries, 0 to 1595862
Data columns (total 20 columns):
date                             1595827 non-null datetime64[ns]
time                             1595827 non-null datetime64[ns]
borough                          1112214 non-null object
zip_code                         1595827 non-null object
latitude                         1595827 non-null float64
longitude                        1595827 non-null float64
persons_injured                  1595827 non-null int64
persons_killed                   1595827 non-null int64
pedestrians_injured              1595827 non-null int64
pedestrians_killed               1595827 non-null int64
cyclist_injured                  1595827 non-null int64
cyclist_killed                   1595827 non-null int64
motorist_injured                 1595827 non-null int64
motorist_killed                  1595827 non-null int64
contributing_factor_vehicle_1    1595827 non-null object
contributing_factor_ve

We've done a lot of data cleaning, and this is a great start for our next stage. We'll save our data locally.

In [70]:
clean_nyc1.to_csv("clean_nyc1_data.csv", index = False)

Now, we can begin our visualizations...

## Data Visualizations

We will create:
* An exploratoy vizualization of fatal crash frequency over time
* An exploratoy vizualization of contributing factors to crash fatalities
* An exploratoy vizualization of fatalities to pedestrians vs cyclists vs motorist
* An exploratoy vizualization of average of deaths by borough
* An exploratoy vizualization of average number of deaths by hour